### ISSUE TITLE
Recognize JDK 10 class format

### ISSUE THREAD

**Issue by melix** [gradle/gradle#3770]
The Java class format has been bumped again for JDK 10. This causes builds to fail. From https://github.com/gradle/gradle/issues/2454#issuecomment-350510963 :

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':util:compileJava'.
> java.lang.IllegalArgumentException (no error message)

Caused by: java.lang.IllegalArgumentException
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:169)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:152)
        at org.gradle.api.internal.tasks.compile.incremental.analyzer.DefaultClassDependenciesAnalyzer.getClassAnalysis(DefaultClassDependenciesAnalyzer.java:32)
```

We should do what we had done for JDK 9, when we were waiting for a release of ASM 6 and that was removed in [this commit](https://github.com/gradle/gradle/commit/e02f4bf8ee705eab8f69559e62c27db9e2e79d37).

@forax given the increased rate of releases of the JDK, the fact that we strongly depend on ASM and knowing that you don't release final versions of ASM until the JDK is out, it would be nice to have an option to ignore this, like our `Java9ClassReader` did, so that we don't have to rely on adding and removing suck hacks again and again.

**Comment by forax** [comment]
This is something we can discuss on the ASM mailing list,
and the support of the JDK 10 in ASM was just merged today:
https://gitlab.ow2.org/asm/asm/commit/ecf4f67cb612e520736ea4e662376fd969570ba5

**Comment by bmuschko** [comment]
@forax Did you already release a version that contains your change?

**Comment by forax** [comment]
Not yet !
The next release, ASM 6.1 BETA, is currently scheduled for the 14th of January,
We still have some work to do.

**Comment by bmuschko** [comment]
@melix I pushed the changes to the branch [`java10-asm-hack`](https://github.com/gradle/gradle/tree/bm/java10-asm-hack). How do I trigger the issue you reported in the issue? Just a regular Java compilation task? We'll need some CI coverage for Java 10 so I opened an [issue](https://github.com/gradle/gradle-private/issues/1066) for it. Maybe we can just wait until ASM 6.1 BETA comes out instead of introducing another hack for this. It likely won't make it into 4.5 though.

**Comment by melix** [comment]
I'm not sure. On the other hand Java 11 previews are already available, so it's likely we'll need something like this soon too because the JavaFX team is using Gradle...

As for reproducing the problem, you need to compile a class using JDK 10, it will have the new class format. Then, I think you need to consume it from another project, so that compile classpath snapshotting kicks in.

**Comment by bmuschko** [comment]
> As for reproducing the problem, you need to compile a class using JDK 10, it will have the new class format. Then, I think you need to consume it from another project, so that compile classpath snapshotting kicks in.

I couldn't quite reproduce the issue. Here's my [sample project](https://github.com/bmuschko/java10-compilation). Any hints would be appreciated!

**Comment by melix** [comment]
@bmuschko see https://github.com/bmuschko/java10-compilation/pull/1

**Comment by bmuschko** [comment]
@melix Thanks for the changes. I opened a [PR with the workaround](https://github.com/gradle/gradle/pull/3941) for review.

**Comment by forax** [comment]
You do need to duplicate the array of bytecode, you can patch the version, create the ClassReader and then patch the array back to its initial value.

**Comment by Godin** [comment]
@forax Hi Remi,

> you can patch the version, create the ClassReader and then **patch the array back to its initial value**

FYI this works only for reading of Java 10 bytecode, but unfortunately won't work for writing/transformation in all cases, because `ClassWriter` invokes `new ClassReader` in case of presence of wide jumps and this invocation will fail on array that contains version `54` - see https://gitlab.ow2.org/asm/asm/blob/ASM_6_0/src/org/objectweb/asm/ClassWriter.java#L1042

**Comment by forax** [comment]
right, it only works when reading the bytecode,
you can always never patch the array back to its initial value or patch it only after all transformations.

**Comment by henrik242** [comment]
Just FYI: ASM 6.1 was just released, and Groovy has already bumped their head revisions to use it.
